mdn/src/main/scala/org/apache/james/mdn/MDNReportParser.scala

```diff
@@ -0,0 +1,475 @@
+      .build()
+    )
+  }
+
+  //    original-message-id-field = "Original-Message-ID" ":" msg-id
+  private[mdn] def originalMessageIdField: Rule1[OriginalMessageId] = rule {
+    "Original-Message-ID" ~ ":" ~ capture(msgId) ~> ((s: String) => new OriginalMessageId(s))
+  }
+
+  //    msg-id          =   [CFWS] "<" id-left "@" id-right ">" [CFWS]
+  private def msgId: Rule0 = rule { optional(cfws) ~ "<" ~ idLeft ~ "@" ~ idRight ~ ">" ~ optional(cfws) }
+
+  //   id-left         =   dot-atom-text / obs-id-left
+  private def idLeft: Rule0 = rule { dotAtomText | obsIdLeft }
+
+  //   obs-id-left     =   local-part
+  private def obsIdLeft: Rule0 = rule { localPart }
+
+  //   obs-id-right    =   domain
+  private def idRight = rule { domain }
+
+  /*    disposition-field =
+                     "Disposition" ":" OWS disposition-mode OWS ";"
+                     OWS disposition-type
+                     [ OWS "/" OWS disposition-modifier
+                     *( OWS "," OWS disposition-modifier ) ] OWS    */
+  private[mdn] def dispositionField : Rule1[Disposition] = rule {
+    ("Disposition" ~ ":" ~ ows ~ dispositionMode ~ ows ~ ";" ~
+    ows ~ dispositionType ~
+    optional(dispositionModifiers) ~ ows) ~> ((modes: (DispositionActionMode, DispositionSendingMode),
+                                                                              dispositionType: DispositionType,
+                                                                              dispositionModifiers: Option[Seq[DispositionModifier]]) =>
+       Disposition.builder()
+         .actionMode(modes._1)
+         .sendingMode(modes._2)
+         .`type`(dispositionType)
+         .addModifiers(dispositionModifiers.getOrElse(Nil):_*)
+         .build()
+      )
+  }
+
+
+
+  //    disposition-mode = action-mode OWS "/" OWS sending-mode
+  private def dispositionMode: Rule1[(DispositionActionMode, DispositionSendingMode)] = rule {
+    (capture(actionMode) ~ ows ~ "/" ~ ows ~ capture(sendingMode)) ~> ((actionMode: String, sendingMode: String) => {
+      val action = actionMode match {
+        case "manual-action" => DispositionActionMode.Manual
+        case "automatic-action" => DispositionActionMode.Automatic
+      }
+      val sending = sendingMode match {
+        case "MDN-sent-manually" => DispositionSendingMode.Manual
+        case "MDN-sent-automatically" => DispositionSendingMode.Automatic
+      }
+      (action, sending)
+    })
+  }
+
+  //    action-mode = "manual-action" / "automatic-action"
+  private def actionMode = rule { "manual-action" | "automatic-action" }
+
+  //    sending-mode = "MDN-sent-manually" / "MDN-sent-automatically"
+  private def sendingMode = rule {"MDN-sent-manually" | "MDN-sent-automatically" }
+
+  /*    disposition-type = "displayed" / "deleted" / "dispatched" /
+                      "processed"    */
+  private def dispositionType : Rule1[DispositionType] = rule {
+    "displayed" ~ push(DispositionType.Displayed) |
```

Comment:
what is this push? it's a little like a capture?

Reply:
It push a value in the value stack. We could have wrote it with a capture but it would have been more verbose.

```
capture("displayed" | "deleted" ...) ~> ((disposition : String) => {
disposition match {
case "displayed" =>  DispositionType.Displayed
case "deleted" =>  DispositionType.Deleted
...
})
```